### PR TITLE
Add Hat/Shoulder/Joystick buttons.  Adjust button geometry.

### PR DESCRIPTION
**Pull request by MisterTea**
Add Hat/Shoulder/Joystick buttons.  Adjust button geometry to make a single control scheme more universal.

The idea now is that face buttons are separated from shoulder/hat buttons.  This removed the weirdness caused by shoulder buttons being BUTTON_3/4 in GBA, BUTTON_5/6 in SNES, and BUTTON_4/5 in n64, and supports the hat-switch-like buttons in n64 and psx to be more naturally mapped.

Also, the button geometry has been changed for 6-button games.  The way the button indices map to geometry is like this:

```
1  4  5
2  3  6
```

While this sounds unorthodox, this mapping is really important, because using this system, 4-button controllers with two shoulder buttons can include 6-button games while still supporting 3-button and 4-button games without changing the button assignment.  If you use something like:

```
4  5  6
1  2  3
```

Then 3/4-button systems like megadrive/snes/psx will use the shoulder button for one of the face buttons, which does not feel as natural as using the 3 face buttons closest to one's thumb.

### CODE REVIEW

**Comment by Tafoid** [comment]
I haven't looked into the changes, but the initial default mapping of the buttons are supposed to make sense for keyboard use which is what most developers use when working with the drivers themselves.  Any enhanced mapping can and has always been done by end users to handle their specific controller(s).  This is just my opinion and wouldn't be surprised if others had the same thoughts.

**Comment by MisterTea** [comment]
@Tafoid It's possible to map the keyboard keys to inputs in a way that makes sense, even with my change.

**Comment by cuavas** [comment]
I really don't think this is a good idea.  It's assuming everyone uses a SNES-derived gamepad as an input device and will just cause confusion for people using keyboard or arcade-style controls.

**Comment by MisterTea** [comment]
@cuavas how so?  I can make the shoulder/hat buttons map to keys on the keyboard by default.  The shoulder/hat buttons are only enabled on console machines, so people using arcade-style controls won't be affected unless they want to play SNES games, in which case they will need to think of a mapping that fits anyways.

**Comment by MisterTea** [comment]
Just to be clear:  This change doesn't change the default mappings (although it might make sense to change them as a result), it adds new inputs to capture buttons that are fundamentally different in nature, such as hat switches and shoulder buttons.

**Comment by etabeta78** [comment]
Typically, new input types got added to MAME/MESS when dozens, if not
hundreds, of supported machines require them (e.g. mahjong inputs, and
later gambling inputs) so that the default mappings can be applied at once
to them all.
Also, this way the emulator offers a reasonable default mapping to users,
making emulation user friendly out-of-the-box.

Based on these considerations, while the addition of the shoulder buttons
and hat inputs might make sense (despite having only a bunch of systems
using them), most of the patch is not really ok because
- by changing the geometry mapping of 6 button games, you make uneasy the
usage with controllers which are not joypads
- the new inputs have no default mappings, so that when you launch e.g. gba
emulation, the shoulder buttons have to be manually configured
- the meaning of JOYSTICK_BUTTON1 & 2 is definitely unclear

In conclusion, I think that having default mappings to standard buttons and
allowing users to remap those to different choices when they have an
appropriate controller capable to better handling hat & shoulder buttons is
way friendlier than the solution proposed by this patch.

**Comment by happppp** [comment]
In cases the d-pad is like N64 C buttons, the method to hook it up in MAME
is a joystick with the flag PORT_16WAY.

some Game & Watch games for example:
https://www.youtube.com/watch?v=K9CoOYqqVLU

**Comment by MisterTea** [comment]
Adding default key bindings to the shoulder buttons is a simple addition to
the PR that I can make.

I disagree that the emulator offers a reasonable default mapping, because
the shoulder buttons map to different keys for each system. With a small
change to this patch, all system shoulder buttons could map to (for
example) 'q' and 'w'.

If the community doesn't like the way I remapped the cps machines, I can
remove that part of the commit and just leave it in MAMEHub.

**Comment by Shideravan** [comment]
I really liked this new approach, as it turn the device input configuration more intuitive for the machine which use these shoulders buttons.
The games/softwares for these machines, in really was developed with this configuration set for shoulder buttons in mind. 
Then I don't see a major problem for users that use keyboards and controllers without shoulders buttons to think the best default way to represent them for their input...

**Comment by happppp** [comment]
Shoulderbuttons: weak-ok from me

Joystick buttons: Besides vage term, it's too rare to be included in the
MAME core. AFAIK of the systems we emulate, this only includes PS1
dual-shock, and the arcade version of (Super?) Punch Out.

hat switch: No, use IPT_JOYSTICK for digital joysticks and d-pads. See my
other email for how to set it up if the directional input is 4 separate
buttons. By the way, don't forget we also have IPT_JOYSTICKLEFT and
IPT_JOYSTICKRIGHT if there is a use case for more than 1 digital joystick.

**Comment by cuavas** [comment]
I still don't buy it for shoulder buttons.  They aren't an inherently different kind of input, they're just another button.  For the same reason on arcade games we don't treat panel-mounted buttons a la Street Fighter as a different input type to joystick triggers like the vulcan gun button in After Burner.

**Comment by MisterTea** [comment]
Mame already treats different buttons as different inputs (notice IPT_START
AND IPT_SELECT). If IPT_START was really a different IPT_BUTTON for each
system, it would be very confusing. Shoulder buttons are currently
confusing for new users.

**Comment by MisterTea** [comment]
Alright so it sounds like shoulder buttons are debatable, and the other changes are out.  This is fine: I can leave them in MAMEHub and remove them from this patch.

When you guys/gals come to a consensus about shoulder/trigger buttons, let me know.

Jason G.

**Comment by rb6502** [comment]
Since there's no consensus on this I'm going to close it for now.

**Comment by MisterTea** [comment]
That's a shame but I kind of expected it given the mame dev reputation.
